tagcwebapp/static/js/main.js

```diff
@@ -0,0 +1,207 @@
+        return {
+            x: x / ratio,
+            y: y / ratio
+        }
+    });
+};
+
+function draw(data, c, translate) {
+    var points = data.cList;
+    var ctx = c.getContext("2d");
+    ctx.clearRect(0, 0, c.width, c.height);
+    $.each(points, function(id, value) {
+        ctx.beginPath();
+        var coor = translate(value.x, value.y);
+
+        ctx.arc(coor.x, coor.y, value.weight / 10, 0, 2 * Math.PI);
+        ctx.stroke();
+        $.each(value.edges, function(key, edge) {
+            if (edge.targetX != -1 && edge.targetY != -1) {
+                ctx.beginPath();
+                ctx.moveTo(coor.x, coor.y);
+                var targetCoor = translate(edge.targetX, edge.targetY);
+                ctx.lineTo(targetCoor.x, targetCoor.y);
+                ctx.lineWidth = edge.weight;
+                ctx.stroke();
+            }
+        });
+    });
+}
+
+function zoom(direction) {
+    var slider = $('.slider', currentHover);
+    var genome = $('#genome');
+    var zoomedGenome = $('#zoomedGenome');
+    var currentWidth = slider.width();
+    var currentHeight = slider.height();
+    var maxWidth = genome.width();
+    var maxHeight = genome.height();
+    var zoomRatio = zoomedGenome.height() / zoomedGenome.width();
+    var newWidth = 0;
+    if (direction > 0) {
+        newWidth = Math.max(Math.min(currentWidth * 0.95, currentWidth - 2), 1);
+    }
+    else {
+        newWidth = Math.min(Math.max(currentWidth * 1.05, currentWidth + 2), maxWidth);
+    }
+    var newHeight = Math.max(currentWidth * zoomRatio, 1);
+    if (newHeight > maxHeight) {
+        newWidth = maxHeight / zoomRatio;
+        newHeight = maxHeight;
+    }
+    var left = parseInt(slider.css('left').replace('px', ''));
+    left = Math.max(0, left + parseInt((currentWidth - newWidth) / 2));
+    if ((left + newWidth) > maxWidth) {
+        left = maxWidth - newWidth;
+    }
+    var top = parseInt(slider.css('top').replace('px', ''));
+    top = Math.max(0, top + parseInt((currentHeight - newHeight) / 2));
+    if ((top + newHeight) > maxHeight) {
+        top = maxHeight - newHeight;
+    }
+    slider.css('left', left + 'px');
+    slider.css('top', top + 'px');
+    slider.width(parseInt(newWidth) +'px');
+    slider.height(parseInt(newHeight) +'px');
+    clearTimeout(zoomTimeout);
+    zoomTimeout = setTimeout(function() { updateZoomedGenome() }, 500);
+}
+
+function updateZoomedGenome()
+{
+    var slider = $('#genome .slider');
+    var x = Math.floor(slider.position().left - slider.parent().position().left);
+    var y = Math.floor(slider.position().top - slider.parent().position().top);
+    var width = slider.width();
+    var height = slider.height();
+    getNodes(x, y, width, height, drawZoom);
+}
+
+function getNodes(x, y, width, height, callback) {
+    $.ajax({
+        url: url + '../getnodes',
+        dataType: 'JSON',
+        type: 'GET',
+        data: {
+            'xleft': x * ratio,
```

Comment:
The way this is done here, is because javascript/ajax is asynchronous, therefor the function that translates it, is passed as a callback function.